### ISSUE TITLE
what about the training results

### ISSUE THREAD

**Issue by taichuai** [johndpope/VASA-1-hack#49]
awesome work, can you show some cases about your inference result

**Comment by johndpope** [comment]
thanks for noticing. 

this model should achieve on par results to the emo volumetric avatar model for inference. (itʼs using it as backbone)
https://github.com/neeek2303/EMOPortraits
this codebase (mit license) includes a fork / nemo (apache) -
 i added a pipeline.py class that can do driving video.

https://drive.google.com/file/d/1PYrUrowldG4roVaJKJb3eEzoLC6IFr8R/view?usp=sharing

when i run this - even on my 5090 - i was only scratching 20 frames / second. 
so to match vasa  -i either need to do some batch processing / cuda torch script - or rip out some computations in that code to get higher throughput. 

the vasa_model is still cooking - i  just replaced the audio to align with joyvasa a few hours ago - and retraining on small set. 
i will run it through (updated) inference in the morning. vi.py - if that fails - i will need to pull apart attention mechanism - (like the expression here)   https://wandb.ai/snoozie/vasa-overfitting/runs/c8bc4sj4?nw=nwusersnoozie

Thereʼs been so many critical errors come out in the wash over last week but thanks to Grok (adding padding to channel ). this caused me so much grief - without the padding - all the training seemed to be working - but when i was running inference - the peoples heads were popping crazy  - it sent me down a rabbit hole. 
https://drive.google.com/file/d/1A726Yq3V9dTPqrYyRUtl6hhD-PcZSZWq/view?usp=sharing
(this was from January)


from emo
<img width="1552" height="790" alt="Image" src="https://github.com/user-attachments/assets/eb1cdc63-e524-4d6d-a423-050b080f3137" />

**Comment by johndpope** [comment]
i threw out most of the vasa_model / trainer code today - and replaced it with some vanilla implementation from grok. seems ok. i did a quick training just then for about 5 mins
https://github.com/user-attachments/assets/492e733b-a41d-4a77-b7d8-8d71f14c8b7b
i will train all night and post results in morning. 


UPDATE - i haven't pushed code yet - but im close to making this all work. 
i overlooked the warping. 

UPDATE - i pushed code to train on the warping - 
https://wandb.ai/snoozie/vasa-overfitting/runs/khyzmhm0?nw=nwusersnoozie

i add this - https://github.com/johndpope/nemo/blob/main/pipeline_face_attr.py
to basically create video from warps - this was succesful today and is now what vasa is aiming to achieve in transformer

for context - i was (incorrectly) just predicting the (rigid warping) theta / translation / rotation - but after days of training - the head movements looked still / flat - i thought the expression embedding was taking care of this - 

https://github.com/user-attachments/assets/e3fb51fa-bf10-4158-a84a-331e600e1b46

After re-reading the paper - we need rigid and non rigid warps.

i add some h5 caching to get all the warps / expressions / emotions / eye gaze angles to speed up training. 
still a work in progress...

**Comment by johndpope** [comment]
getting closer - 
this is critical clean warp extraction from nemo - 
https://github.com/johndpope/VASA-1-hack/commit/4fb7904020294cf8c3e4748bfe8f940ed59bc042

<img width="1121" height="531" alt="Image" src="https://github.com/user-attachments/assets/c84d8c5b-c7bf-470b-a954-5e23c0fd011b" />

 ive isolated the warps successfully - with caches to h5 file - so given a video + image - can reapply the warps - this will help in fps in vasa - i just gotta wire up the dataset to extract the warps from video -

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/ef238541-3d55-4436-8f0d-d7b359b2e809" />

TODO - go through the training videos - > extract the warps - store them - cache h5 file
wire up the diffusion motion transformer to predict warps - 
connect the decoder.

**Comment by johndpope** [comment]
<img width="768" height="256" alt="Image" src="https://github.com/user-attachments/assets/6a7f291b-8135-484a-ad0f-00c36691c5a8" />

https://wandb.ai/snoozie/vasa-overfitting/runs/9d0di1su

**Comment by johndpope** [comment]
https://wandb.ai/snoozie/vasa-overfitting/runs/bsl6966l?nw=nwusersnoozie
I throw in a cross attention into decoder layers for audio - 
finally results are coming good. 

<img width="1457" height="564" alt="Image" src="https://github.com/user-attachments/assets/3f6bb81c-27c3-4420-8b01-92a83800b776" />

the overfitting is target 2 layer depth - so it trains quickly.
i turn off a bunch of losses - just to isolate things.
i am overfitting to one video - will expand tomorrow.


some progress - had to rejig model.
https://wandb.ai/snoozie/vasa-overfitting/runs/qqprlbs6/panel/b0gp2dkme?nw=nwusersnoozie

**Comment by johndpope** [comment]
after a day of training on 5090 - it was crap - i added a cheat sheet db for expressions - 

https://github.com/johndpope/VASA-1-hack/pull/61
it solves the expression floating in space - like this 

<img width="1632" height="378" alt="Image" src="https://github.com/user-attachments/assets/13021236-a2bc-4751-824f-4d54a147bca9" />

step 0 !!!!
<img width="2117" height="1701" alt="Image" src="https://github.com/user-attachments/assets/469a543b-9d19-4026-be8c-85e29f6f267d" />
wont need 4x a6000 training if i can sample a huge dataset of expressions.

[vasa-output-overfit_config-videovideoeI2V8Bd5X9s-scene6_scene1.mp4
](https://drive.google.com/file/d/12CJqxsE50Xpk00yZvFF82jwnJJor5BBL/view?usp=drive_link)

**Comment by johndpope** [comment]
https://drive.google.com/drive/u/0/folders/1kmERS5AGQbGPbvb2t6OIIH3h7PMtidbC

https://drive.google.com/file/d/1DQ0CcH93m5rrcC2VCMXs9_PyhPzrZBUR/view?usp=drive_link
it's a little bit better - but now looking at lip sync losses - any ideas - im all ears. i guess the synchformer may help....

im overfitting to a training video - but it's not aligning. ...


UPDATE - 
my latest commit implements the syncformer loss correctly - 
the gradients blew up mid training - but i just need a good video training set.
